<!--![Banner](https://github.com/microbiome/microbiome/blob/master/vignettes/figure/composition-example4-1.png)-->
<!--[![Follow](https://img.shields.io/twitter/follow/ropengov.svg?style=social)](https://twitter.com/intent/follow?screen_name=ropengov)-->


microbiome R package
==========

[![Build Status](https://api.travis-ci.org/microbiome/microbiome.png)](https://travis-ci.org/microbiome/microbiome)
[![codecov.io](https://codecov.io/github/microbiome/microbiome/coverage.svg?branch=master)](https://codecov.io/github/microbiome/microbiome?branch=master)  


The microbiome R package provides tools for the exploration and
analysis of microbiome profiling data. It extends the
[phyloseq](https://github.com/joey711/phyloseq) framework in R. 
Your feedback, suggestions and bug reports are welcome.


### Installation and use

See the package [tutorial](http://microbiome.github.io/microbiome/).


### Contribute

Contributions are very welcome:

  * [Issue Tracker](https://github.com/microbiome/microbiome/issues) 
  * [Pull requests](https://github.com/microbiome/microbiome/)
  * Subscribe to the [mailing list](https://groups.google.com/forum/#!forum/microbiome-devel) (microbiome-devel@googlegroups.com)
  * [Star us on the Github page](https://github.com/microbiome/microbiome)


### Publications using the microbiome package

**Kindly cite this work** as follows: "Leo Lahti [et al.](https://github.com/microbiome/microbiome/graphs/contributors) (2017). microbiome: Tools for microbiome analysis. R package version <[version number here](DESCRIPTION)>. http://microbiome.github.com", or citing the references below when appropriate (different references introduce or take advantage of some of these methods). The list of publications is not exhaustive. Let us know if you know of further publications using the microbiome package; we are collecting these on the website.

[Intestinal microbiome landscaping: Insight in community assemblage and implications for microbial modulation strategies](https://academic.oup.com/femsre/article/doi/10.1093/femsre/fuw045/2979411/Intestinal-microbiome-landscaping-insight-in#58802539). Sudarshan Shetty, Floor Hugenholtz, Leo Lahti, Hauke Smidt, Willem M de Vos, Antoine Danchin. _FEMS Microbiology Reviews_ fuw045, 2017.

[Metagenomics meets time series analysis: unraveling microbial community dynamics](http://dx.doi.org/10.1016/j.mib.2015.04.004) Karoline Faust, Leo Lahti, Didier Gonze, Willem M de Vos, Jeroen Raes. _Current Opinion in Microbiology_ 15:56-66 2015.

[Tipping elements in the human intestinal ecosystem](http://www.nature.com/ncomms/2014/140708/ncomms5344/full/ncomms5344.html) Leo Lahti, Jarkko Salojärvi, Anne Salonen, Marten Scheffer, Willem M. de Vos. _Nature Communications_ 5:4344, 2014. 

[Fat, Fiber and Cancer Risk in African, Americans and Rural Africans](http://www.nature.com/ncomms/2015/150428/ncomms7342/full/ncomms7342.html)  Stephen O’Keefe, Jia V. Li, Leo Lahti, Junhai Ou, Franck
Carbonero, Khaled Mohammed, Joram M. Posma, James Kinross, Elaine
Wahl, Elizabeth Ruder, Kishore Vipperla, Vasudevan Naidoo, Lungile
Mtshali, Sebastian Tims, Philippe G. B. Puylaert, James DeLany, Alyssa
Krasinskas, Ann C. Benefiel, Hatem O. Kaseb, Keith Newton, Jeremy K.
Nicholson, Willem M. de Vos, H. Rex Gaskins, Erwin G. Zoetendal.
_Nature Communications_ 6:6342, 2015.

[Associations between the human intestinal microbiota, Lactobacillus rhamnosus GG and serum lipids indicated by integrated analysis of high-throughput profiling data](http://dx.doi.org/10.7717/peerj.32) Leo Lahti, Anne Salonen, Riina A. Kekkonen, Jarkko Salojärvi, Jonna Jalanka-Tuovinen, Airi Palva, Matej Orešič, Willem M. de Vos. _PeerJ_ 1:e32, 2013.

[The adult intestinal core microbiota is determined by analysis depth and health status](http://onlinelibrary.wiley.com/doi/10.1111/j.1469-0691.2012.03855.x/abstract) Anne Salonen, Jarkko Salojärvi, Leo Lahti, and Willem M. de Vos. _Clinical Microbiology and Infection_ 18(S4):16 20, 2012. 


### Acknowledgements

Author: [Leo Lahti](https://github.com/antagomir/)

[Contributors](https://github.com/microbiome/microbiome/graphs/contributors): Jarkko Salojärvi, Tineka Blake. 

Financial support has been provided by the following bodies:

  * Academy of Finland (grants 256950 and 295741)
  * [University of Turku](http://www.utu.fi/en/Pages/home.aspx), Department of Mathematics and Statistics
  * [VIB lab for Bioinformatics and (eco-)systems biology](http://www.vib.be/en/research/scientists/Pages/Jeroen-Raes-Lab.aspx), VIB/KULeuven, Belgium
  * [Molecular Ecology group](http://www.mib.wur.nl/UK/), Laboratory of Microbiology, Wageningen University, Netherlands
  * [Department of Veterinary Bioscience](http://www.vetmed.helsinki.fi/apalva/index.htm), University of Helsinki, Finland.

This work critically relies on the independent [phyloseq](https://github.com/joey711/phyloseq) package and data structures for R-based microbiome analysis developed by Paul McMurdie and Susan Holmes. 



